drivers: hwinfo: Add drivers for NXP SoC

alexanderwachter opened the pull request:
This PR adds hwinfo device id driver support for NXP LPC and kinetis SOC.

Commit:
drivers: hwinfo: add driver support for kinetis device ID

Add driver support for kinetis ID device.
The length depends on the SoC.
`SIM_GetUniqueId` was not used because the struct would reorder
the ID and makes the driver more complicated because the length
of the struct depends on the SoC.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

alexanderwachter commented:
This drivers are not tested. Looking for someone who can test them since i have no NXP boards.
Tests:
cmake -DBOARD=lpcxpresso54114_m0 ../../tests/drivers/hwinfo/api
cmake -DBOARD=frdm_k64f ../../tests/drivers/hwinfo/api

MaureenHelm commented on drivers/hwinfo/hwinfo_mcux_kinetis.c:
```diff
@@ -0,0 +1,54 @@
+	dev_id.id[2] = SIM->UIDMH;
+#if defined(SIM_UIDH)
+	dev_id.id[3] = SIM->UIDH;
+#endif /* SIM_UIDH */
```
Rather than duplicating all the ifdef logic, why not just call `SIM_GetUniqueId`?

alexanderwachter replied:
It's more complicated because the struct is also modified by SIM_UIDH ...

MaureenHelm replied:
You're already copying the id twice. Why can't you do this:
```c
ssize_t _impl_hwinfo_get_device_id(u8_t *buffer, size_t length)
{
        sim_uid_t sim_uid;
        SIM_GetUniqueId(&sim_uid);                                                         

        if (length > sizeof(sim_uid)) {
                length = sizeof(sim_uid);
        }

        memcpy(buffer, &sim_uid, length);

        return length;
}
```

alexanderwachter replied:
Because of the reordering of high and low words.

MaureenHelm replied:
I'm not sure that matters, since it's still unique if you rearrange the words. Will you please add a note in the commit message explaining why you did not use SIM_GetUniqueId?

alexanderwachter replied:
Added the note

MaureenHelm reviewed (changes requested):
Please add `hwinfo` to the supported list for frdm_k64f and lpcxpresso54114_m4 to ensure these drivers get built in CI. I tested both on hardware and they pass the test, however the lpcxpresso54114_m4 returns a unique ID of 0xFFFFFFFF which is suspicious. @galak, @stanislav-poboril 	could you please test this on your lpc board?

MaureenHelm commented:
> was the return value positive?

Yes. It suggests that the special area of flash where this value is stored has been erased on my particular board, hence my question to @galak and @stanislav-poboril to check their boards.

Commit:
drivers: hwinfo: add driver support for NXP i.mx RT device ID

Add driver support for NXP i.mx RT ID device.
This device has an ID of 8 bytes.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

stanislav-poboril commented:
> 
> 
> Please add `hwinfo` to the supported list for frdm_k64f and lpcxpresso54114_m4 to ensure these drivers get built in CI. I tested both on hardware and they pass the test, however the lpcxpresso54114_m4 returns a unique ID of 0xFFFFFFFF which is suspicious. @galak, @stanislav-poboril could you please test this on your lpc board?

I tested on my board and got `0xFFFF0000000000000000000000000000`.

alexanderwachter commented:
Resolved the conflicts

alexanderwachter commented:
Removed LPC driver because of unreliable results.

MaureenHelm commented:
I tested the imx rt driver on mimxrt1050_evk and it works well.

otavio commented:
@chtavares592 please use this PR for testing when reworking #13039 to use `hwinfo` API.
